Implementation of `instanceof` operator

morrien opened the pull request:
- fixed @@hasinstance symbol call
- added `get_method()` and `ordinary_has_instance()` for `GcObject`
- implemented `instanceof` operator

This Pull Request fixes/closes #796.

HalidOdat commented on boa/src/object/gcobject.rs:
```diff
@@ -676,6 +676,73 @@ impl GcObject {
+                Ok(false)
+            } else {
+                Err(context.construct_type_error(format!(
+                    "{}",
+                    "function has non-object prototype in instanceof check"
```
```suggestion
                    "function has non-object prototype in instanceof check",
```
We dont have to have a place holder for the main string we can just inline it

HalidOdat commented on boa/src/syntax/ast/node/operator/bin_op/mod.rs:
```diff
@@ -125,15 +125,29 @@ impl Executable for BinOp {
+                                }
+                                None => {
+                                    return interpreter.throw_type_error(format!(
+                                        "{}",
+                                        "right-hand side of 'instanceof' is not callable"
```
```suggestion
                                        "right-hand side of 'instanceof' is not callable",
```
same here

HalidOdat reviewed (commented):
Looks good! this PR needs a rebase, and we should also add some tests :)

morrien commented:
- fixed `throw` formatting
- added tests for `instanceof` operator and `ordinary_has_instance()`
- rebased the PR

@HalidOdat could you please explain how to resolve the remaining failing checks? Thank you.

Razican commented:
>     * fixed `throw` formatting
> 
>     * added tests for `instanceof` operator and `ordinary_has_instance()`
> 
>     * rebased the PR
> 
> 
> @HalidOdat could you please explain how to resolve the remaining failing checks? Thank you.

It seems that this requires running `cargo fmt`. I will review the code, since it also seems to require a small change to fix clippy lints. The test262 issue should be solved in #914.

Razican commented on boa/src/object/gcobject.rs:
```diff
@@ -676,6 +676,71 @@ impl GcObject {
+                ))
+            }
+        } else {
+            return Ok(false);	
```
This should solve the clippy warning:
```suggestion
            Ok(false)
```

Razican commented on boa/src/object/tests.rs:
```diff
@@ -0,0 +1,16 @@
+        "#;
+
+    assert_eq!(&exec(scenario), "\"TypeError: function has non-object prototype in instanceof check\"");
+}
```
```suggestion
}

```

Razican commented on boa/src/syntax/ast/node/operator/tests.rs:
```diff
@@ -26,3 +26,31 @@ fn assignmentoperator_rhs_throws_error() {
+        "#;
+
+    assert_eq!(&exec(scenario), "\"TypeError: right-hand side of 'instanceof' is not callable\"");
+}
```
```suggestion
}

```

Razican reviewed (commented):
I added some suggestions on new lines at the end of files. This should be added by any editor that takes the editorconfig into account, but it seems it didn't happen in this case.

morrien commented:
I ran `cargo fmt`, which also added the EOF newlines, and I fixed the clippy warning sa well.
I'll wait for [#914](https://github.com/boa-dev/boa/pull/914) to be merged and rebase the PR again.

The only failed check that remains is code coverage, how could I address that? Thank you.

Razican commented:
> I ran `cargo fmt`, which also added the EOF newlines, and I fixed the clippy warning sa well.
> I'll wait for [#914](https://github.com/boa-dev/boa/pull/914) to be merged and rebase the PR again.

Thanks! That one is ready now :)

> The only failed check that remains is code coverage, how could I address that? Thank you.

We are currently debating about "patch" coverage failures. We will merge it anyways if that's the only issue.

morrien commented:
Latest rebase seems to have fixed everything, even somehow improved the code coverage - I believe this is ready for merge.

With that being said, I had to rebase twice, because running `cargo clippy` would not detect another clippy warning that I noticed only in the "Files changed" tab afterwards - does anyone know what could be causing this? I believe this caused me to miss the first clippy warning regarding `return Ok(false)` as well, as I only noticed it after the PR went online.

Razican commented:
> With that being said, I had to rebase twice, because running `cargo clippy` would not detect another clippy warning that I noticed only in the "Files changed" tab afterwards - does anyone know what could be causing this? I believe this caused me to miss the first clippy warning regarding `return Ok(false)` as well, as I only noticed it after the PR went online.

This might be related to https://github.com/rust-lang/rust-clippy/issues/4612

Razican reviewed (approved):
Looks good to me, thanks!

### Test262 conformance changes:
  | Test result | master count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 78,415 | 78,415 | 0 |
  | Passed | 18,461 | 18,945 | **+484** |
  | Ignored | 15,547 | 15,547 | 0 |
  | Failed | 44,407 | 43,923 | **-484** |
  | Panics | 1,129 | 1,127 | **-2** |
  | Conformance | 23.54 | 24.16 | **+0.62%** |
